keps/sig-scheduling/201900715-node-teardown-pressure.md

```diff
@@ -0,0 +1,173 @@
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+For the current implementation of the kubernetes scheduler the teardown phase of the container, which may consist of removing a large amount of files, is considered free. This may result in a buildup of containers in termination phase which will render a node unusable due to io congestion.
+
+This KEP proposes a teardown pressure condition that is triggered when the termination load reaches a critical threshold signaling the scheduler to spare the node any additional workload until the backlog is resolved.
+
+Additionally, to help distribute load better, a TeardownPriority classifier is suggested, that negatively weighs the number of 
+containers under removal into the node weight.
+
+## Motivation
+
+While trying to stabilize a very particular workload, we where able to find a corner case in the kubernetes scheduler that can cause runaway overload of one or more nodes, seriously impacting the cluster performance.
+The corner case very much depends on the specific hardware setup and is not always easy triggerable, but in general results in a buildup of containers in termination state on one or multiple nodes until they become unresponsive.
+
+  1. After starting of the container, download a very large archive with 10ks of files.
+  2. Do something very minor, short lived
+  3. Exit
+
+The container then enters the termination phase in which docker removes the files from the overlay filesystem. If the removal process is to slow to finish before next container enters the termination phase, a buildup of those containers can occure.
+We found while stressing a [elasticio] workload, that some nodes accumulated 70 containers in termination state, rendering them unusable. Kubernetes-scheduler still scheduled the pods on this node while other nodes had very little load.
```

Comment:
There are other alternatives to resolve this issue for many people, such as spreading the workload more strongly.  The scheduler is generally not aware of startup or shutdown resource use in general, which would lead to better decisions than can be done on shutdown alone.  Startup pressure is much more common than shutdown but shares many similarities (jvms / dbs / caches that must warm up).  An approach that tried to better estimate these costs would be more broadly applicable than the current proposal.

Reply:
Yes, most containers need more resources on startup, that's true, but when a container is started, it's resources are already accounted for and the usual resource borrowing of the linux scheduler takes place. Of course this can be optimized, but our case is different. It's unaccounted accumulation of used resources that causes a cluster to crash, which as far as I understood should not happen with kubernetes. In fact, this workload makes also problems on AWS and other cloud providers. When the massive start and termination of containers causes a kubernetes cluster to get into non usable state, there is something seriously broken. We only use the label matching for getting this task workload on the designated machines, that this setting will cause such unbalanced workload can hardly be the designated behavior.